compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -1041,7 +1041,39 @@ impl<'a> Linker for EmLinker<'a> {
         &mut self.cmd
     }
 
-    fn set_output_kind(&mut self, _output_kind: LinkOutputKind, _out_filename: &Path) {}
+    fn set_output_kind(&mut self, output_kind: LinkOutputKind, _out_filename: &Path) {
+        match output_kind {
+            LinkOutputKind::DynamicNoPicExe | LinkOutputKind::DynamicPicExe => {
+                // "-sMAIN_MODULE=1" breaks
+                // https://github.com/rust-lang/rust/issues/92738
+                self.cmd.arg("-sMAIN_MODULE=2");
+                warn!(
+                    "Building dynamic executable with -sMAIN_MODULE=2. \
+                    Dead code elimination may break things. \
+                    See https://emscripten.org/docs/compiling/Dynamic-Linking.html?highlight=main_module#code-size \
+                    "
+                );
```

Comment:
I think so. @sbc100 

Maybe it would be a good idea to switch to using `StaticPicExe` by default, since that matches the expectations of the Emscripten linker better. I guess to do that we would want to set `crt_static_default` to `true` and `crt_static_respected` to `true`? This might be a slightly disruptive change though. At the very least it would be good to set `crt_static_respected` to `true`

Generally if you pass `-sMAIN_MODULE=2` then you should also be passing all the side modules you are going to load, for instance:
```
emcc a1.o a2.o a3.o -o a.so -sSIDE_MODULE=2
emcc b1.o b2.o b3.o -o b.so -sSIDE_MODULE=2
emcc m1.o m2.o m3.o a.so b.so -o main.js -sMAIN_MODULE=2
```
then loading `a.so` and `b.so` is guaranteed to work, but `main.js` only contains support for runtime features used in `a.so` and `b.so` so if you try to load `c.so` and it uses some feature not used in `a.so` and `b.so` it will break. That's what this is warning about. Passing `-sEXPORT_ALL` makes the situation a bit better but there are some things that work with `-sMAIN_MODULE=1` that break with `-sMAIN_MODULE=2 -sEXPORT_ALL`.

We should ideally use `-sMAIN_MODULE=1` here but we can't because it wraps the linking step with `--whole-archive`, something like: 
```
clang --whole-archive some_rust_lib.rlib -o main
```
but this doesn't work because it says `wasm-ld: error: unknown file type: lib.rmeta`.

Reply:
Though using `crt-static` to toggle dynamic linking support is weird because the *only* option is statically linking the C runtime, it cannot be dynamically linked. The question is whether we want to support dynamically linking other things.

Reply:
I see, yes, if you are enabling `-fPIC` when `StaticPicExe` makes sense.

Reply:
I don't mean to say that Emscripten actually generates a PIC executable, just that the Rust output kind is going to be `StaticPicExe` not `StaticNoPicExe`.